spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java

```diff
@@ -254,31 +269,51 @@ public void removeTopic(String... topic) {
 		}
 	}
 
-	private synchronized void connectAndSubscribe() throws MqttException { // NOSONAR
+	private synchronized void connect() throws MqttException { // NOSONAR
 		MqttConnectOptions connectionOptions = this.clientFactory.getConnectionOptions();
 		this.cleanSession = connectionOptions.isCleanSession();
 		this.consumerStopAction = this.clientFactory.getConsumerStopAction();
 		if (this.consumerStopAction == null) {
 			this.consumerStopAction = ConsumerStopAction.UNSUBSCRIBE_CLEAN;
 		}
-		Assert.state(getUrl() != null || connectionOptions.getServerURIs() != null,
-				"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
-		this.client = this.clientFactory.getClientInstance(getUrl(), getClientId());
-		this.client.setCallback(this);
-		if (this.client instanceof MqttClient) {
-			((MqttClient) this.client).setTimeToWait(getCompletionTimeout());
+
+		long completionTimeout = getCompletionTimeout();
+		var clientManager = getClientManager();
+		if (clientManager == null) {
+			Assert.state(getUrl() != null || connectionOptions.getServerURIs() != null,
+					"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
+			this.client = this.clientFactory.getAsyncClientInstance(getUrl(), getClientId());
+			this.client.setCallback(this);
+			this.client.connect(connectionOptions).waitForCompletion(completionTimeout);
+			this.client.setManualAcks(isManualAcks());
+		}
+		else {
+			this.client = clientManager.getClient();
+		}
+	}
+
+	private void subscribe() {
+		var clientManager = getClientManager();
+		if (clientManager != null && this.client == null) {
+			this.client = clientManager.getClient();
 		}
 
 		this.topicLock.lock();
 		String[] topics = getTopic();
 		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
 		try {
-			this.client.connect(connectionOptions);
-			this.client.setManualAcks(isManualAcks());
 			if (topics.length > 0) {
 				int[] requestedQos = getQos();
-				int[] grantedQos = Arrays.copyOf(requestedQos, requestedQos.length);
-				this.client.subscribe(topics, grantedQos);
+				IMqttMessageListener listener = this::messageArrived;
+				IMqttMessageListener[] listeners = Stream.of(topics)
+						.map(t -> listener)
+						.toArray(IMqttMessageListener[]::new);
+				IMqttToken subscribeToken = this.client.subscribe(topics, requestedQos, listeners);
+				subscribeToken.waitForCompletion(getCompletionTimeout());
+				int[] grantedQos = subscribeToken.getGrantedQos();
+				if (grantedQos.length == 1 && grantedQos[0] == 0x80) {
```

Comment:
Magic number? `// NOSONAR` ?

Reply:
This is an interesting case 🙂 
Initially v3 adapter had a blocking client impl and I have changed it to use async client.
One test case started failing after that, it was `MqttAdapterTests#testSubscribeFailure`. 
You can see it mocks this magic number `0x80`:
`given(token.getGrantedQos()).willReturn(new int[]{ 0x80 });`

I have looked through the code of a blocking client and it is just a wrapper around async one - but it really has this strange difference in `subscribe` method, `/org/eclipse/paho/client/mqttv3/MqttClient.java:465`
```java
		if (grantedQos.length == 1 && qos[0] == 0x80) {
			throw new MqttException(MqttException.REASON_CODE_SUBSCRIBE_FAILED);
		}
```
So I decided to keep the behavior, especially given we had a unit test for it.

I will add `//NOSONAR` for now..

Reply:
Sure! Makes sense. Let's live with this for now only in v3 impl!